templates/fine-tune-stable-diffusion/README.ipynb

```diff
@@ -0,0 +1,392 @@
+    "* `ray.init(runtime_env={\"env_vars\": ...})` sets the environment variables on all workers in the cluster -- setting the environment variable in this notebook on the head node is not enough in a distributed setting.\n",
+    "* `train_fn_per_worker` is the function that will run on all distributed training workers. In this case, it's just a light wrapper on top of the `diffusers` example script that copies the latest checkpoint to shared cluster storage.\n",
+    "* `ScalingConfig` is the configuration that determines how many workers and what kind of accelerator to use for training. Once the training is launched, **Anyscale will automatically scale up nodes to meet this resource request!**\n",
+    "\n",
+    "The result of this fine-tuning will be a fine-tuned LoRA model checkpoint at `MODEL_CHECKPOINT_PATH`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "MODEL_CHECKPOINT_PATH = os.environ[\"ANYSCALE_ARTIFACT_STORAGE\"] + \"/checkpoint-final\"\n",
+    "\n",
+    "print(\"Final checkpoint will be uploaded to: \", MODEL_CHECKPOINT_PATH)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "\n",
+    "import ray\n",
+    "import ray.train\n",
+    "from ray.train.torch import TorchTrainer\n",
+    "\n",
+    "from train_dreambooth_lora_sdxl import main\n",
+    "from utils import download_from_cloud, upload_to_cloud\n",
+    "\n",
+    "\n",
+    "# Set environment variables across the entire cluster.\n",
+    "WANDB_API_KEY = os.environ.get(\"WANDB_API_KEY\")\n",
+    "if WANDB_API_KEY:\n",
+    "    ray.init(\n",
+    "        runtime_env={\"env_vars\": {\"WANDB_API_KEY\": WANDB_API_KEY}},\n",
+    "        ignore_reinit_error=True,\n",
+    "    )\n",
+    "\n",
+    "\n",
+    "def train_fn_per_worker(config: dict):\n",
+    "    download_from_cloud(cloud_uri=DATA_CLOUD_PATH, local_path=SUBJECT_IMAGES_PATH)\n",
+    "\n",
+    "    # See train_dreambooth_lora_sdxl.py for all of the training details.\n",
+    "    final_checkpoint_path = main(config[\"args\"])\n",
+    "\n",
+    "    # Upload final checkpoint to cloud. (Only the rank 0 worker will return a path here.)\n",
+    "    if final_checkpoint_path is not None:\n",
+    "        upload_to_cloud(local_path=final_checkpoint_path, cloud_uri=MODEL_CHECKPOINT_PATH)\n",
+    "        print(\"Final checkpoint has been uploaded to: \", MODEL_CHECKPOINT_PATH)\n",
+    "\n",
+    "\n",
+    "trainer = TorchTrainer(\n",
+    "    train_fn_per_worker,\n",
+    "    # Pass command line arguments to the `config` dict of the `train_fn_per_worker`\n",
+    "    train_loop_config={\"args\": TRAINING_ARGS},\n",
+    "    scaling_config=ray.train.ScalingConfig(\n",
+    "        # Do data parallel training with A10G GPU workers\n",
+    "        num_workers=4, use_gpu=True, accelerator_type=\"A10G\"\n",
```

Comment:
Problem with this:
* A10G is not available on GCP clouds, only L4
* No easy way to detect which cloud is being used.
* Not possible to specify minimum GPU memory requirement.

Possible solutions:
* Janky way of detecting cloud by checking the `ANYSCALE_ARTIFACT_STORAGE`
* Add a helper that does [requests instance metadata](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html) and determines the cloud that way.
* Ask the user to uncomment code that switches the accelerator type to L4 if they're running on GCP.

Reply:
Yeah just add a helper in utils to `get_A10G_class_gpu()` or similar.